examples/filters_example.cpp

```diff
@@ -0,0 +1,327 @@
+                        timer.stop();
+                    });
+                }
+            }
+
+            if (!progress_str.empty())
+                ImGui::BulletText(progress_str.c_str());
+
+            ImGui::ProgressBar(recon_progress, ImVec2(0.f, 0.f));
+            if (execution_handle.valid() && execution_handle.wait_for(std::chrono::microseconds(
+                                                0u)) == std::future_status::ready)
+            {
+                execution_handle.get();
+                viewer.data().clear();
+                auto const V = from_point_cloud(points);
+                viewer.data().add_points(V, Eigen::RowVector3d(1.0, 1.0, 0.0));
+                viewer.data().point_size = 1.f;
+                viewer.core().align_camera_center(V);
+                auto const duration =
+                    std::chrono::duration_cast<std::chrono::milliseconds>(timer.ops.front().second);
+                timer.ops.clear();
+                progress_str = "Execution time: " + std::to_string(duration.count()) + " ms";
+            }
+
+            ImGui::TreePop();
+        }
+        ImGui::End();
+    };
+
+    viewer.core().set_rotation_type(igl::opengl::ViewerCore::RotationType::ROTATION_TYPE_TRACKBALL);
+    viewer.data().show_lines = false;
+    viewer.launch();
+
+    return 0;
+}
+void apply_filter(std::vector<pcp::point_t>& points, std::atomic<float>& progress)
+{
+    auto const coordinate_map = [](pcp::point_t const& p) {
+        return std::array<float, 3u>{p.x(), p.y(), p.z()};
+    };
+    progress                           = 0.f;
+    float constexpr num_ops            = 4.f;
+    float constexpr progress_increment = 1.f / 2.f;
+    auto const progress_forward        = [&]() {
+        progress = progress + progress_increment;
+    };
+
+    using kdtree_type = pcp::basic_linked_kdtree_t<pcp::point_t, 3u, decltype(coordinate_map)>;
+    pcp::kdtree::construction_params_t params{};
+    params.max_depth    = 4u;
+    params.construction = pcp::kdtree::construction_t::nth_element;
+    pcp::statistical_outlier_filter::construction_params_t<float> statistical_filter_params{};
+    pcp::pass_through_filter::construction_params_t<float> pass_through_filter_params{};
+    pcp::radius_outlier_filter::construction_params_t<float> radius_outlier_filter_params{};
+    pcp::density_filter::construction_params_t<float> density_filter_params{};
+
+    kdtree_type kdtree{points.begin(), points.end(), coordinate_map, params};
+    progress_forward();
+    auto const knn_map = [&](pcp::point_t const& p) {
+        std::size_t num_neighbors = static_cast<std::size_t>(mean_k);
+        return kdtree.nearest_neighbours(p, num_neighbors);
+    };
+    auto const point_map = [](pcp::point_t const& p) {
+        return p;
+    };
+    auto const range_search_map = [&](pcp::point_t p, float radius) {
+        pcp::sphere_a<float> ball;
+        ball.position[0] = coordinate_map(p)[0];
```

Comment:
Extract the coordinates into a local variable and then access through [i], please. Each call to coordinate map might create a new object, so just call it once.